flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveFunctionDefinitionFactory.java

```diff
@@ -0,0 +1,149 @@
+
+	private final String hiveVersion;
+
+	public HiveFunctionDefinitionFactory(HiveConf hiveConf) {
+		checkNotNull(hiveConf);
+
+		this.hiveVersion = new JobConf(hiveConf).get(HiveCatalogValidator.CATALOG_HIVE_VERSION, HiveShimLoader.getHiveVersion());
+	}
+
+	@Override
+	public FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction) {
+
+		String functionClassName = catalogFunction.getClassName();
+
+		if (Boolean.valueOf(catalogFunction.getProperties().get(CatalogConfig.IS_GENERIC))) {
+			throw new TableException(
+				String.format("HiveFunctionDefinitionFactory does not support generic functions %s yet", name));
+		}
+
+		Class clazz;
+		try {
+			clazz = Thread.currentThread().getContextClassLoader().loadClass(functionClassName);
+
+			LOG.info("Successfully loaded Hive udf '{}' with class '{}'", name, functionClassName);
+		} catch (ClassNotFoundException e) {
+			throw new TableException(
+				String.format("Failed to initiate an instance of class %s.", functionClassName), e);
+		}
+
+		if (UDF.class.isAssignableFrom(clazz)) {
+			LOG.info("Transforming Hive function '{}' into a HiveSimpleUDF", name);
+
+			return new ScalarFunctionDefinition(
+				name,
+				new HiveSimpleUDF(new HiveFunctionWrapper<>(functionClassName))
+			);
+		} else if (GenericUDF.class.isAssignableFrom(clazz)) {
+			LOG.info("Transforming Hive function '{}' into a HiveGenericUDF", name);
+
+			return new ScalarFunctionDefinition(
+				name,
+				new HiveGenericUDF(new HiveFunctionWrapper<>(functionClassName))
+			);
+		} else if (GenericUDTF.class.isAssignableFrom(clazz)) {
+			LOG.info("Transforming Hive function '{}' into a HiveGenericUDTF", name);
+
+			HiveGenericUDTF udtf = new HiveGenericUDTF(new HiveFunctionWrapper<>(functionClassName));
+
+			return new TableFunctionDefinition(
+				name,
+				udtf,
+				GenericTypeInfo.of(Row.class)
+			);
+		} else if (GenericUDAFResolver2.class.isAssignableFrom(clazz)) {
+			LOG.info(
+				"Transforming Hive function '{}' into a HiveGenericUDAF with no UDAF bridging and Hive version %s",
+				name, hiveVersion);
+
+			HiveGenericUDAF udaf = new HiveGenericUDAF(new HiveFunctionWrapper<>(functionClassName), false, hiveVersion);
+
+			return new AggregateFunctionDefinition(
+				name,
+				udaf,
+				GenericTypeInfo.of(Object.class),
+				GenericTypeInfo.of(GenericUDAFEvaluator.AggregationBuffer.class)
+			);
+		} else if (UDAF.class.isAssignableFrom(clazz)) {
+			LOG.info(
+				"Transforming Hive function '{}' into a HiveGenericUDAF with UDAF bridging and Hive version %s",
+				name, hiveVersion);
+
+			HiveGenericUDAF udaf = new HiveGenericUDAF(new HiveFunctionWrapper<>(functionClassName), true, hiveVersion);
+
+			return new AggregateFunctionDefinition(
+				name,
+				udaf,
+				GenericTypeInfo.of(Object.class),
+				GenericTypeInfo.of(GenericUDAFEvaluator.AggregationBuffer.class)
+			);
+		} else {
+			throw new IllegalArgumentException(
```

Comment:
Minor: Is there any possible to register flink function to hive catalog?

Reply:
Timo and I decided to limit the scope of function catalog features in 1.9.0 to be just reading hive functions. Please see our discussions above.